drivers: ethernet: Add VIRTIO Network Device

fkokosinski opened the pull request:
This PR adds a driver for the VIRTIO Ethernet device. This is a minimal driver which sets a MAC address and transmits packets, but does not support any extra features like the control channel or checksum offloading.

Commit:
drivers: ethernet: Add VIRTIO Network Device

Add a driver for the VIRTIO Ethernet device.
This is a minimal driver which sets a MAC address and transmits packets,
but does not support any extra features like the control channel or
checksum offloading.
Confirmed to work with the networking subsystem samples. For example, the
zperf sample shows a result of 85 Mbps download and 14 Mbps upload.

Signed-off-by: Jakub Klimczak <jklimczak@internships.antmicro.com>

Commit:
samples: net: echo_server: Add files for VIRTIO network device

Create an overlay and a config file for testing the VIRTIO network device.
Add some instructions to the README.rst file as well.
These can be used with any of the networking subsystem samples.

Signed-off-by: Jakub Klimczak <jklimczak@internships.antmicro.com>

maass-hamburg commented on drivers/ethernet/eth_virtio_net.c:
```diff
@@ -0,0 +1,254 @@
+	const struct virtnet_config *config = dev->config;
+	struct virtnet_data *data = dev->data;
+
+#if DT_INST_PROP(0, zephyr_random_mac_address)
```
this driver is instance based, access these values in VIRTIO_NET_DEFINE

local_mac_address can directly go in the data struct, zephyr_random_mac_address should be a bool in config, that is checked here

maass-hamburg reviewed (dismissed):
LFTM as a minimal implementation

pdgendt reviewed (commented):
Maybe add a subsection in the sample's README.rst on how to use this?

jukkar commented:
There is also https://docs.zephyrproject.org/latest/connectivity/networking/networking_with_host.html that has various options for host connectivity. Perhaps add information there how to use this driver?

Commit:
tests: drivers: build_all: virtio: Add VIRTIO Network device

Add the VIRTIO Network device to the driver build_all test on
qemu_x86_64 (PCI) and qemu_cortex_a53 (MMIO). Additionally,
restrict the test to these two platforms, getting rid of qemu_x86,
as enabling networking would cause build errors on many boards,
including the qemu_x86. Use dashes in node names.

Signed-off-by: Jakub Klimczak <jklimczak@internships.antmicro.com>

Commit:
doc: connectivity: networking: Mention VIRTIO Network device in doc

Add instructions on running echo_server using the VIRTIO Network device
in the QEMU Ethernet documentation.

Signed-off-by: Jakub Klimczak <jklimczak@internships.antmicro.com>

fkokosinski commented on tests/drivers/build_all/virtio/boards/qemu_cortex_a53.overlay:
```diff
@@ -13,8 +13,14 @@
 	virtio_entropy: virtio_entropy {
 		compatible = "virtio,device4";
 		status = "okay";
+		zephyr,random-mac-address;
```
Ooops, thanks for the catch. Should be fixed now

pdgendt commented on tests/drivers/build_all/virtio/boards/qemu_cortex_a53.overlay:
```diff
@@ -18,3 +18,13 @@
+&virtio_mmio1 {
+	status = "okay";
+
+	virtio_net: virtio_net {
```
[Node names](https://docs.zephyrproject.org/latest/contribute/style/devicetree.html) should use dashes instead of underscores (might want to fix `virtio_entropy` while you're at it 🫣 

```suggestion
	virtio_net: virtio-net {
```

pdgendt commented on drivers/ethernet/Kconfig.virtio_net:
```diff
@@ -0,0 +1,22 @@
+
+if ETH_VIRTIO_NET
+
+config ETH_NIC_MODEL
+	string
+	default "virtio"
```
Is this supposed to be prompt-less? How can a user update this value?

pdgendt commented on drivers/ethernet/eth_virtio_net.c:
```diff
@@ -0,0 +1,259 @@
+	return 0;
+}
+
+static struct ethernet_api virtnet_api = {.iface_api.init = virtnet_if_init,
+					  .get_capabilities = virtnet_get_capabilities,
+					  .send = virtnet_send};
```
Prefer using a trailing comma so fields are formatted after a single tab character

```suggestion
static struct ethernet_api virtnet_api = {
	.iface_api.init = virtnet_if_init,
	.get_capabilities = virtnet_get_capabilities,
	.send = virtnet_send,
};
```

pdgendt commented on drivers/ethernet/eth_virtio_net.c:
```diff
@@ -0,0 +1,261 @@
+
+	if (pkt == NULL) {
+		LOG_ERR("received packet, but could not pass it to the operating system");
+	} else if (net_pkt_write(pkt, &(data->rxb[buf_no][sizeof(struct _virtio_net_hdr)]), len)) {
+		LOG_ERR("could not copy entire received packet");
+		net_pkt_unref(pkt);
+	} else if (net_recv_data(data->iface, pkt)) {
```
```suggestion
	} else if (net_pkt_write(pkt, &(data->rxb[buf_no][sizeof(struct _virtio_net_hdr)]), len) < 0) {
		LOG_ERR("could not copy entire received packet");
		net_pkt_unref(pkt);
	} else if (net_recv_data(data->iface, pkt) < 0) {
```

pdgendt commented on drivers/ethernet/eth_virtio_net.c:
```diff
@@ -0,0 +1,261 @@
+	if (data->virtio_devcfg == NULL) {
+		LOG_ERR("could not get config struct");
+	}
+	if (virtio_commit_feature_bits(config->vdev)) {
```
```suggestion
	if (virtio_commit_feature_bits(config->vdev) < 0) {
```

pdgendt commented on drivers/ethernet/eth_virtio_net.c:
```diff
@@ -0,0 +1,261 @@
+	struct virtq_buf vqbuf[] = {
+		{.addr = data->txb, .len = sizeof(struct _virtio_net_hdr) + len}};
+
+	if (virtq_add_buffer_chain(vq, vqbuf, 1, 1, NULL, NULL, K_FOREVER)) {
```
```suggestion
	if (virtq_add_buffer_chain(vq, vqbuf, 1, 1, NULL, NULL, K_FOREVER) < 0) {
```

pdgendt commented on drivers/ethernet/eth_virtio_net.c:
```diff
@@ -0,0 +1,261 @@
+	struct virtnet_data *data = dev->data;
+	size_t len = net_pkt_get_len(pkt);
+
+	if (net_pkt_read(pkt, data->txb + sizeof(struct _virtio_net_hdr), len)) {
```
```suggestion
	if (net_pkt_read(pkt, data->txb + sizeof(struct _virtio_net_hdr), len) < 0) {
```

pdgendt reviewed (commented):
Nitpicks, still implicit boolean conversions throughout the code.

As a recommendation I'd advise to adopt using `int ret;` where possible to provide the user with some useful runtime feedback.

```c
int ret;

ret = some_func();
if (ret < 0) {
	LOG_ERR("some_func failed (%d)", ret);
	return ret;
}
